File: flutter/nepali_date_picker_kmp/android/src/main/kotlin/dev/shivathapaa/nepali_date_picker_kmp/Mapping.kt
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

package dev.shivathapaa.nepali_date_picker_kmp

import dev.shivathapaa.nepalidatepickerkmp.NepaliSelectableDates
import dev.shivathapaa.nepalidatepickerkmp.calendar_model.NepaliDateConverter
import dev.shivathapaa.nepalidatepickerkmp.data.CustomCalendar
import dev.shivathapaa.nepalidatepickerkmp.data.CustomDateTime
import dev.shivathapaa.nepalidatepickerkmp.data.DigitScript
import dev.shivathapaa.nepalidatepickerkmp.data.MonthCalendar
import dev.shivathapaa.nepalidatepickerkmp.data.NameFormat
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliDateFormatStyle
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliDateFormatter
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliDateLocale
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliDatePickerLang
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliMonthCalendar
import dev.shivathapaa.nepalidatepickerkmp.data.SimpleDate
import dev.shivathapaa.nepalidatepickerkmp.data.SimpleTime
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliCalendarEvent
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliCalendarPolicy
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliDayStatus
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliEventKind
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliEventProvider
import dev.shivathapaa.nepalidatepickerkmp.event.excludingClosures
import dev.shivathapaa.nepalidatepickerkmp.event.excludingWeekends

// Converters between the pigeon wire types and the shared engine's types.
// Enum wire values map by ordinal; both declaration orders are the contract.

internal fun LangDto.toCore(): NepaliDatePickerLang = NepaliDatePickerLang.entries[ordinal]

internal fun NameFormatDto.toCore(): NameFormat = NameFormat.entries[ordinal]

internal fun DateFormatStyleDto.toCore(): NepaliDateFormatStyle =
    NepaliDateFormatStyle.entries[ordinal]

internal fun DigitScriptDto.toCore(): DigitScript = DigitScript.entries[ordinal]

internal fun EventKindDto.toCore(): NepaliEventKind = NepaliEventKind.entries[ordinal]

internal fun NepaliEventKind.toDto(): EventKindDto = EventKindDto.entries[ordinal]

internal fun DatePatternDto.toCore(): NepaliDateFormatter.Pattern =
    NepaliDateFormatter.Pattern.entries[ordinal]

internal fun LocaleDto.toCore(): NepaliDateLocale = NepaliDateLocale(
    language = language.toCore(),
    dateFormat = dateFormat.toCore(),
    weekDayName = weekDayName.toCore(),
    monthName = monthName.toCore(),
    digitScript = digitScript?.toCore()
)

internal fun DateDto.toCore(): SimpleDate =
    SimpleDate(year.toInt(), month.toInt(), dayOfMonth.toInt())

internal fun TimeDto.toCore(): SimpleTime =
    SimpleTime(hour.toInt(), minute.toInt(), second.toInt(), nanosecond.toInt())

internal fun SimpleTime.toDto(): TimeDto = TimeDto(
    hour = hour.toLong(),
    minute = minute.toLong(),
    second = second.toLong(),
    nanosecond = nanosecond.toLong()
)

internal fun CalendarDto.toCore(): CustomCalendar = CustomCalendar(
    year = year.toInt(),
    month = month.toInt(),
    dayOfMonth = dayOfMonth.toInt(),
    era = era.toInt(),
    firstDayOfMonth = firstDayOfMonth.toInt(),
    lastDayOfMonth = lastDayOfMonth.toInt(),
    totalDaysInMonth = totalDaysInMonth.toInt(),
    dayOfWeekInMonth = dayOfWeekInMonth.toInt(),
    dayOfWeek = dayOfWeek.toInt(),
    dayOfYear = dayOfYear.toInt(),
    weekOfMonth = weekOfMonth.toInt(),
    weekOfYear = weekOfYear.toInt()
)

internal fun CustomCalendar.toDto(): CalendarDto = CalendarDto(
    year = year.toLong(),
    month = month.toLong(),
    dayOfMonth = dayOfMonth.toLong(),
    era = era.toLong(),
    firstDayOfMonth = firstDayOfMonth.toLong(),
    lastDayOfMonth = lastDayOfMonth.toLong(),
    totalDaysInMonth = totalDaysInMonth.toLong(),
    dayOfWeekInMonth = dayOfWeekInMonth.toLong(),
    dayOfWeek = dayOfWeek.toLong(),
    dayOfYear = dayOfYear.toLong(),
    weekOfMonth = weekOfMonth.toLong(),
    weekOfYear = weekOfYear.toLong()
)

internal fun NepaliMonthCalendar.toDto(): MonthInfoDto = MonthInfoDto(
    year = year.toLong(),
    month = month.toLong(),
    totalDaysInMonth = totalDaysInMonth.toLong(),
    firstDayOfMonth = firstDayOfMonth.toLong(),
    lastDayOfMonth = lastDayOfMonth.toLong(),
    daysFromStartOfWeekToFirstOfMonth = daysFromStartOfWeekToFirstOfMonth.toLong()
)

internal fun MonthCalendar.toDto(): MonthInfoDto = MonthInfoDto(
    year = year.toLong(),
    month = month.toLong(),
    totalDaysInMonth = totalDaysInMonth.toLong(),
    firstDayOfMonth = firstDayOfMonth.toLong(),
    lastDayOfMonth = lastDayOfMonth.toLong(),
    daysFromStartOfWeekToFirstOfMonth = daysFromStartOfWeekToFirstOfMonth.toLong()
)

internal fun CustomDateTime.toDto(): DateTimeDto =
    DateTimeDto(calendar = customCalendar.toDto(), time = simpleTime.toDto())

internal fun IntRange.toDto(): YearRangeDto =
    YearRangeDto(first = first.toLong(), last = last.toLong())

internal fun EventDto.toCore(): NepaliCalendarEvent = NepaliCalendarEvent(
    date = SimpleDate(year.toInt(), month.toInt(), dayOfMonth.toInt()),
    name = name,
    kind = kind.toCore(),
    closesOffices = closesOffices,
    id = id,
    payload = payload
)

internal fun NepaliCalendarEvent.toDto(): EventDto = EventDto(
    year = date.year.toLong(),
    month = date.month.toLong(),
    dayOfMonth = date.dayOfMonth.toLong(),
    name = name,
    kind = kind.toDto(),
    closesOffices = closesOffices,
    id = id,
    payload = payload
)

internal fun NepaliDayStatus.toDto(): DayStatusDto = DayStatusDto(
    isWeeklyOff = isWeeklyOff,
    isNonWorking = isNonWorking,
    primaryKind = primaryKind?.toDto(),
    names = names,
    events = events.map { it.toDto() },
    closures = closures.map { it.toDto() }
)

/**
 * A provider over a fixed table, equal to another one built from the same
 * events, so remembered policies stay stable across identical calls.
 */
internal data class TableEventProvider(
    private val byYear: Map<Int, Set<NepaliCalendarEvent>>
) : NepaliEventProvider {
    override fun events(year: Int): Set<NepaliCalendarEvent> = byYear[year].orEmpty()
}

internal fun PolicyDto.toCore(): NepaliCalendarPolicy = NepaliCalendarPolicy(
    weeklyOffDays = weeklyOffDays.map { it.toInt() }.toSet(),
    provider = TableEventProvider(
        events.map { it.toCore() }
            .groupBy { it.date.year }
            .mapValues { (_, list) -> list.toSet() }
    )
)

/**
 * The engine's selectable rule this description composes, or null when every
 * date is selectable.
 */
internal fun SelectableDto?.toCore(): NepaliSelectableDates? {
    if (this == null) return null
    val min = minDate?.toCore()
    val max = maxDate?.toCore()
    var selectable: NepaliSelectableDates? = when {
        min != null && max != null -> NepaliDateConverter.DateRangeSelectable(
            minDate = min,
            maxDate = max,
            includeMinDate = includeMinDate,
            includeMaxDate = includeMaxDate
        )
        min != null -> NepaliDateConverter.AfterDateSelectable(min, includeMinDate)
        max != null -> NepaliDateConverter.BeforeDateSelectable(max, includeMaxDate)
        else -> null
    }
    val weekend = excludeWeekend?.map { it.toInt() }?.toSet()
    val closures = excludeClosuresOf?.toCore()
    if (weekend != null || closures != null) {
        var composed = selectable ?: AllDatesSelectable
        if (weekend != null) composed = composed.excludingWeekends(weekend)
        if (closures != null) composed = composed.excludingClosures(closures.provider)
        selectable = composed
    }
    return selectable
}

/** Every date selectable, the interface's own defaults. */
internal object AllDatesSelectable : NepaliSelectableDates

File: flutter/nepali_date_picker_kmp/ios/nepali_date_picker_kmp/Sources/nepali_date_picker_kmp/Mapping.swift
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

// Converters between the pigeon wire types and the shared engine's types.
// Enum wire values map by ordinal; both declaration orders are the contract.

import Foundation
import nepali_date_picker

extension LangDto {
  var toCore: NepaliDatePickerLang { self == .english ? .english : .nepali }
}

extension NameFormatDto {
  var toCore: NameFormat {
    switch self {
    case .full: return .full
    case .medium: return .medium
    case .short: return .short_
    }
  }
}

extension DateFormatStyleDto {
  var toCore: NepaliDateFormatStyle {
    switch self {
    case .full: return .full
    case .long: return .long_
    case .medium: return .medium
    case .shortMdy: return .shortMdy
    case .shortYmd: return .shortYmd
    case .compactMdy: return .compactMdy
    case .compactYmd: return .compactYmd
    }
  }
}

extension DigitScriptDto {
  var toCore: DigitScript { self == .latin ? .latin : .devanagari }
}

extension EventKindDto {
  var toCore: NepaliEventKind {
    switch self {
    case .governmentPublic: return .governmentpublic
    case .religious: return .religious
    case .regional: return .regional
    case .observance: return .observance
    }
  }
}

extension NepaliEventKind {
  var toDto: EventKindDto {
    switch self {
    case .governmentpublic: return .governmentPublic
    case .religious: return .religious
    case .regional: return .regional
    default: return .observance
    }
  }
}

extension DatePatternDto {
  var toCore: NepaliDateFormatter.Pattern {
    switch self {
    case .yyyySlashMmSlashDd: return .yyyySlashMmSlashDd
    case .yyyyDashMmDashDd: return .yyyyDashMmDashDd
    case .ddSlashMmSlashYyyy: return .ddSlashMmSlashYyyy
    case .ddDashMmDashYyyy: return .ddDashMmDashYyyy
    }
  }
}

extension LocaleDto {
  var toCore: NepaliDateLocale {
    NepaliDateLocale(
      language: language.toCore,
      dateFormat: dateFormat.toCore,
      weekDayName: weekDayName.toCore,
      monthName: monthName.toCore,
      digitScript: digitScript?.toCore
    )
  }
}

extension DateDto {
  var toCore: SimpleDate {
    SimpleDate(year: Int32(year), month: Int32(month), dayOfMonth: Int32(dayOfMonth))
  }
}

extension TimeDto {
  var toCore: SimpleTime {
    SimpleTime(
      hour: Int32(hour),
      minute: Int32(minute),
      second: Int32(second),
      nanosecond: Int32(nanosecond)
    )
  }
}

extension SimpleTime {
  var toDto: TimeDto {
    TimeDto(
      hour: Int64(hour),
      minute: Int64(minute),
      second: Int64(second),
      nanosecond: Int64(nanosecond)
    )
  }
}

extension SimpleDate {
  var toDateDto: DateDto {
    DateDto(year: Int64(year), month: Int64(month), dayOfMonth: Int64(dayOfMonth))
  }
}

extension CalendarDto {
  var toCore: CustomCalendar {
    CustomCalendar(
      year: Int32(year),
      month: Int32(month),
      dayOfMonth: Int32(dayOfMonth),
      era: Int32(era),
      firstDayOfMonth: Int32(firstDayOfMonth),
      lastDayOfMonth: Int32(lastDayOfMonth),
      totalDaysInMonth: Int32(totalDaysInMonth),
      dayOfWeekInMonth: Int32(dayOfWeekInMonth),
      dayOfWeek: Int32(dayOfWeek),
      dayOfYear: Int32(dayOfYear),
      weekOfMonth: Int32(weekOfMonth),
      weekOfYear: Int32(weekOfYear)
    )
  }
}

extension CustomCalendar {
  var toDto: CalendarDto {
    CalendarDto(
      year: Int64(year),
      month: Int64(month),
      dayOfMonth: Int64(dayOfMonth),
      era: Int64(era),
      firstDayOfMonth: Int64(firstDayOfMonth),
      lastDayOfMonth: Int64(lastDayOfMonth),
      totalDaysInMonth: Int64(totalDaysInMonth),
      dayOfWeekInMonth: Int64(dayOfWeekInMonth),
      dayOfWeek: Int64(dayOfWeek),
      dayOfYear: Int64(dayOfYear),
      weekOfMonth: Int64(weekOfMonth),
      weekOfYear: Int64(weekOfYear)
    )
  }
}

extension NepaliMonthCalendar {
  var toDto: MonthInfoDto {
    MonthInfoDto(
      year: Int64(year),
      month: Int64(month),
      totalDaysInMonth: Int64(totalDaysInMonth),
      firstDayOfMonth: Int64(firstDayOfMonth),
      lastDayOfMonth: Int64(lastDayOfMonth),
      daysFromStartOfWeekToFirstOfMonth: Int64(daysFromStartOfWeekToFirstOfMonth)
    )
  }
}

extension MonthCalendar {
  var toDto: MonthInfoDto {
    MonthInfoDto(
      year: Int64(year),
      month: Int64(month),
      totalDaysInMonth: Int64(totalDaysInMonth),
      firstDayOfMonth: Int64(firstDayOfMonth),
      lastDayOfMonth: Int64(lastDayOfMonth),
      daysFromStartOfWeekToFirstOfMonth: Int64(daysFromStartOfWeekToFirstOfMonth)
    )
  }
}

extension CustomDateTime {
  var toDto: DateTimeDto {
    DateTimeDto(calendar: customCalendar.toDto, time: simpleTime.toDto)
  }
}

extension KotlinIntRange {
  var toDto: YearRangeDto { YearRangeDto(first: Int64(first), last: Int64(last)) }
}

extension EventDto {
  var toCore: NepaliCalendarEvent {
    NepaliCalendarEvent(
      date: SimpleDate(year: Int32(year), month: Int32(month), dayOfMonth: Int32(dayOfMonth)),
      name: name,
      kind: kind.toCore,
      closesOffices: closesOffices,
      id: id,
      payload: payload
    )
  }
}

extension NepaliCalendarEvent {
  var toDto: EventDto {
    EventDto(
      year: Int64(date.year),
      month: Int64(date.month),
      dayOfMonth: Int64(date.dayOfMonth),
      name: name,
      kind: kind.toDto,
      closesOffices: closesOffices,
      id: id,
      payload: payload
    )
  }
}

extension NepaliDayStatus {
  var toDto: DayStatusDto {
    DayStatusDto(
      isWeeklyOff: isWeeklyOff,
      isNonWorking: isNonWorking,
      primaryKind: primaryKind?.toDto,
      names: names,
      events: events.map { $0.toDto },
      closures: closures.map { $0.toDto }
    )
  }
}

/// A provider over a fixed table; equal tables compare equal on the Kotlin
/// side, keeping remembered policies stable across identical calls.
final class TableEventProvider: NSObject, NepaliEventProvider {
  private let byYear: [Int32: Set<NepaliCalendarEvent>]

  init(events: [NepaliCalendarEvent]) {
    var table: [Int32: Set<NepaliCalendarEvent>] = [:]
    for event in events {
      table[event.date.year, default: []].insert(event)
    }
    byYear = table
    super.init()
  }

  func events(year: Int32) -> Set<NepaliCalendarEvent> { byYear[year] ?? [] }

  func closesOn(date: SimpleDate) -> Bool {
    events(year: date.year).contains { $0.date == date && $0.closesOffices }
  }
}

extension PolicyDto {
  var toCore: NepaliCalendarPolicy {
    NepaliCalendarPolicy(
      weeklyOffDays: Set(weeklyOffDays.map { KotlinInt(int: Int32($0)) }),
      provider: TableEventProvider(events: events.map { $0.toCore })
    )
  }
}

/// The engine's selectable rule this description composes, or nil when
/// every date is selectable.
func selectableFromDto(_ dto: SelectableDto?) -> NepaliSelectableDates? {
  guard let dto else { return nil }
  let converter = NepaliDateConverter.shared
  var selectable: NepaliSelectableDates?
  if let min = dto.minDate?.toCore, let max = dto.maxDate?.toCore {
    selectable = converter.DateRangeSelectable(
      minDate: min,
      maxDate: max,
      includeMinDate: dto.includeMinDate,
      includeMaxDate: dto.includeMaxDate
    )
  } else if let min = dto.minDate?.toCore {
    selectable = converter.AfterDateSelectable(simpleDate: min, includeDate: dto.includeMinDate)
  } else if let max = dto.maxDate?.toCore {
    selectable = converter.BeforeDateSelectable(simpleDate: max, includeDate: dto.includeMaxDate)
  }
  let weekend = dto.excludeWeekend.map { Set($0.map { KotlinInt(int: Int32($0)) }) }
  let closures = dto.excludeClosuresOf?.toCore
  if weekend != nil || closures != nil {
    var composed = selectable ?? NepaliDatePickerDefaults.shared.AllDates
    if let weekend {
      composed = EventHelpersKt.excludingWeekends(composed, weekend: weekend)
    }
    if let closures {
      composed = EventHelpersKt.excludingClosures(composed, provider: closures.provider)
    }
    selectable = composed
  }
  return selectable
}

extension EventOptionsDto {
  var toIos: NepaliEventOptions {
    let options = NepaliEventOptions()
    options.weeklyOffDays = weeklyOffDays.map { KotlinInt(int: Int32($0)) }
    options.events = events.map { mark in
      NepaliEventInfo(
        year: Int32(mark.year),
        month: Int32(mark.month),
        dayOfMonth: Int32(mark.dayOfMonth),
        name: mark.name,
        kind: mark.kind.toCore,
        closesOffices: mark.closesOffices,
        colorArgb: Int32(truncatingIfNeeded: mark.colorArgb),
        indicate: mark.indicate
      )
    }
    options.markWeeklyOff = markWeeklyOff
    options.markEvents = markEvents
    options.tintContainer = tintContainer
    options.indicateWeeklyOff = indicateWeeklyOff
    options.describeEvents = describeEvents
    options.weeklyOffColorArgb = Int32(truncatingIfNeeded: weeklyOffColorArgb)
    options.publicHolidayColorArgb = Int32(truncatingIfNeeded: publicHolidayColorArgb)
    options.religiousColorArgb = Int32(truncatingIfNeeded: religiousColorArgb)
    options.regionalColorArgb = Int32(truncatingIfNeeded: regionalColorArgb)
    options.observanceColorArgb = Int32(truncatingIfNeeded: observanceColorArgb)
    options.markedContainerColorArgb = Int32(truncatingIfNeeded: markedContainerColorArgb)
    return options
  }
}

func calendarSystemOfEra(_ era: Int64) -> CalendarSystem {
  CalendarSystem.companion.fromEra(era: Int32(era)) ?? .bikramSambat
}
